File: clients/JavaScript/5.0 mcp_store_user/MCPStoreUserClient.js
const net = require('net');
const readline = require('readline');
const { argv, exit } = require('process');

/**
 * Funktion zum Parsen der Kommandozeilenargumente
 * @param {string[]} args - Array von Kommandozeilenargumenten
 * @returns {Object} - Objekt mit geparsten Argumenten
 */
function parseArguments(args) {
    const parsedArgs = {};
    for (let i = 2; i < args.length; i++) {
        switch (args[i]) {
            case '--server-ip':
                if (i + 1 < args.length) {
                    parsedArgs.serverIp = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --server-ip angegeben.');
                }
                break;
            case '--server-port':
                if (i + 1 < args.length) {
                    parsedArgs.serverPort = parseInt(args[++i], 10);
                } else {
                    console.warn('⚠️ Kein Wert für --server-port angegeben.');
                }
                break;
            case '--token':
                if (i + 1 < args.length) {
                    parsedArgs.token = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --token angegeben.');
                }
                break;
            case '--name':
                if (i + 1 < args.length) {
                    parsedArgs.name = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --name angegeben.');
                }
                break;
            case '--email':
                if (i + 1 < args.length) {
                    parsedArgs.email = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --email angegeben.');
                }
                break;
            case '--password':
                if (i + 1 < args.length) {
                    parsedArgs.password = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --password angegeben.');
                }
                break;
            case '--language':
                if (i + 1 < args.length) {
                    parsedArgs.language = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --language angegeben.');
                }
                break;
            case '--timezone':
                if (i + 1 < args.length) {
                    parsedArgs.timezone = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --timezone angegeben.');
                }
                break;
            case '--roles':
                parsedArgs.roles = [];
                while (i + 1 < args.length && !args[i + 1].startsWith('--')) {
                    parsedArgs.roles.push(args[++i]);
                }
                break;
            case '--groups':
                parsedArgs.groups = [];
                while (i + 1 < args.length && !args[i + 1].startsWith('--')) {
                    parsedArgs.groups.push(args[++i]);
                }
                break;
            case '--usePublic':
                parsedArgs.usePublic = true;
                break;
            case '--activateFtp':
                parsedArgs.activateFtp = true;
                break;
            case '--ftpPassword':
                if (i + 1 < args.length) {
                    parsedArgs.ftpPassword = args[++i];
                } else {
                    console.warn('⚠️ Kein Wert für --ftpPassword angegeben.');
                }
                break;
            default:
                console.warn(`⚠️ Unbekanntes Argument: ${args[i]}`);
        }
    }
    return parsedArgs;
}

/**
 * Funktion zum interaktiven Abfragen eines Parameters (optional)
 * @param {string} query - Frage an den Benutzer
 * @returns {Promise<string>} - Antwort des Benutzers
 */
function askQuestion(query) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: true
    });

    return new Promise((resolve) => {
        rl.question(query, (answer) => {
            rl.close();
            resolve(answer);
        });
    });
}

/**
 * Sendet eine Anfrage an den MCP-Server, um einen neuen Benutzer zu erstellen.
 *
 * @param {string} serverIp - IP-Adresse des MCP-Servers
 * @param {number} serverPort - Portnummer des MCP-Servers
 * @param {string} token - Authentifizierungstoken
 * @param {string} name - Name des neuen Benutzers
 * @param {string} email - Email des neuen Benutzers
 * @param {string} password - Passwort für den neuen Benutzer
 * @param {string} language - Bevorzugte Sprache des neuen Benutzers
 * @param {string} timezone - Zeitzone des neuen Benutzers
 * @param {string[]} roles - Rollen des neuen Benutzers
 * @param {string[]} groups - Gruppen des neuen Benutzers
 * @param {boolean} usePublic - Verwendung der öffentlichen Wissensbasis
 * @param {boolean} activateFtp - Aktivierung von FTP für den Benutzer
 * @param {string} ftpPassword - FTP-Passwort für den Benutzer
 * @returns {Promise<Object>} - Antwort vom Server
 */
function sendStoreUserRequest(serverIp, serverPort, token, name, email, password, language, timezone, roles, groups, usePublic, activateFtp, ftpPassword) {
    return new Promise((resolve, reject) => {
        const client = new net.Socket();
        const payload = {
            command: "store_user",
            token: token,
            arguments: {
                name: name,
                email: email,
                password: password,
                language: language,
                timezone: timezone,
                roles: roles,
                groups: groups,
                usePublic: usePublic,
                activateFtp: activateFtp,
                ftpPassword: ftpPassword
            }
        };
        const payloadString = JSON.stringify(payload);

        // Timeout setzen (optional)
        const TIMEOUT_DURATION = 10000; // 10 Sekunden
        const timeout = setTimeout(() => {
            client.destroy(); // Verbindung zerstören
            reject(new Error('Verbindungs-Timeout: Der Server hat nicht rechtzeitig geantwortet.'));
        }, TIMEOUT_DURATION);

        client.connect(serverPort, serverIp, () => {
            console.log(`🔗 Verbindung zum Server (${serverIp}:${serverPort}) hergestellt.`);
            console.log(`📤 Sende Payload: ${payloadString}`);
            client.write(payloadString);
        });

        let responseData = '';

        client.on('data', (data) => {
            console.log(`📥 Empfangene Daten: ${data}`);
            responseData += data.toString();
            try {
                const parsedData = JSON.parse(responseData);
                console.log('✅ JSON-Antwort erfolgreich geparst.');
                clearTimeout(timeout);
                resolve(parsedData);
                client.destroy(); // Verbindung schließen
            } catch (err) {
                console.warn('⚠️ Antwort noch nicht vollständig oder ungültiges JSON. Weitere Daten werden erwartet.');
                // Weiter empfangen
            }
        });

        client.on('close', () => {
            console.log('🔒 Verbindung zum Server geschlossen.');
            clearTimeout(timeout);
        });

        client.on('error', (err) => {
            console.error('❌ Verbindungsfehler:', err.message);
            clearTimeout(timeout);
            reject(err);
        });
    });
}

// Hauptfunktion
async function main() {
    const args = argv;
    const parsedArgs = parseArguments(args);
    let { 
        serverIp, 
        serverPort, 
        token, 
        name, 
        email, 
        password, 
        language, 
        timezone, 
        roles, 
        groups, 
        usePublic, 
        activateFtp, 
        ftpPassword 
    } = parsedArgs;

    // Überprüfen, ob alle erforderlichen Parameter vorhanden sind, sonst interaktiv abfragen
    if (!serverIp) {
        serverIp = await askQuestion('🔗 Bitte gib die Server-IP ein: ');
    }
    if (!serverPort) {
        const portInput = await askQuestion('🔗 Bitte gib den Server-Port ein: ');
        serverPort = parseInt(portInput, 10);
    }
    if (!token) {
        token = await askQuestion('🔒 Bitte gib dein Authentifizierungstoken ein: ');
    }
    if (!name) {
        name = await askQuestion('👤 Bitte gib den Namen des Benutzers ein: ');
    }
    if (!email) {
        email = await askQuestion('📧 Bitte gib die Email des Benutzers ein: ');
    }
    if (!password) {
        password = await askQuestion('🔑 Bitte gib das Passwort des Benutzers ein: ');
    }
    if (!language) {
        language = await askQuestion('🌐 Bitte gib die bevorzugte Sprache des Benutzers ein (z.B. en, de): ');
    }
    if (!timezone) {
        timezone = await askQuestion('🕰️ Bitte gib die Zeitzone des Benutzers ein (z.B. Europe/Berlin): ');
    }
    // Rollen und Gruppen sind optional und wurden bereits mit parseArguments behandelt
    // usePublic, activateFtp und ftpPassword sind ebenfalls optional

    // Standardwerte für optionale Parameter setzen, falls sie nicht vorhanden sind
    roles = roles || [];
    groups = groups || [];
    usePublic = usePublic || false;
    activateFtp = activateFtp || false;
    ftpPassword = ftpPassword || '';

    try {
        console.log('🧑‍💻 Sende Store-User-Anfrage...');
        const response = await sendStoreUserRequest(
            serverIp,
            serverPort,
            token,
            name,
            email,
            password,
            language,
            timezone,
            roles,
            groups,
            usePublic,
            activateFtp,
            ftpPassword
        );
        console.log('✔️ Antwort vom Server:', JSON.stringify(response, null, 2));
    } catch (err) {
        console.error('❌ Fehler:', err.message);
    }
}

main();

File: clients/C++/5.0 mcp_store_user/MCPStoreUserClient.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <sstream>
#include <stdexcept>
#include <json/json.h>
#include <winsock2.h>
#include <ws2tcpip.h>

#pragma comment(lib, "ws2_32.lib") // Verlinkung mit der Winsock-Bibliothek

// Funktion zum Parsen von Argumenten
std::map<std::string, std::string> parseArguments(int argc, char* argv[]) {
    std::map<std::string, std::string> args;
    for (int i = 1; i < argc; ++i) {
        std::string key = argv[i];
        if (i + 1 < argc && key.rfind("--", 0) == 0) {
            args[key] = argv[++i];
        }
    }
    return args;
}

// Funktion zum Parsen von Listen-Argumenten
std::vector<std::string> parseListArgument(int argc, char* argv[], const std::string& key) {
    std::vector<std::string> values;
    for (int i = 1; i < argc; ++i) {
        if (argv[i] == key && i + 1 < argc) {
            for (int j = i + 1; j < argc && std::string(argv[j]).rfind("--", 0) != 0; ++j) {
                values.push_back(argv[j]);
            }
        }
    }
    return values;
}

// Funktion zum Senden der Anfrage
std::string sendRequest(const std::string& serverIp, int serverPort, const Json::Value& payload) {
    Json::StreamWriterBuilder writer;
    std::string payloadJson = Json::writeString(writer, payload);

    // Winsock initialisieren
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        throw std::runtime_error("Failed to initialize Winsock.");
    }

    // Socket erstellen
    SOCKET sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET) {
        WSACleanup();
        throw std::runtime_error("Failed to create socket.");
    }

    // Server-Adresse konfigurieren
    sockaddr_in serverAddr;
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(serverPort);
    if (inet_pton(AF_INET, serverIp.c_str(), &serverAddr.sin_addr) <= 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Invalid server IP address.");
    }

    // Verbindung herstellen
    if (connect(sock, (sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Connection failed.");
    }

    // Daten senden
    if (send(sock, payloadJson.c_str(), payloadJson.size(), 0) < 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Failed to send data.");
    }

    // Antwort empfangen
    char buffer[4096];
    int bytesRead;
    std::ostringstream response;

    do {
        bytesRead = recv(sock, buffer, sizeof(buffer) - 1, 0);
        if (bytesRead > 0) {
            buffer[bytesRead] = '\0'; // Null-terminieren
            response << buffer;
        }
    } while (bytesRead == sizeof(buffer) - 1);

    // Socket schließen
    closesocket(sock);
    WSACleanup();

    return response.str();
}

int main(int argc, char* argv[]) {
    try {
        auto args = parseArguments(argc, argv);

        // Argumente auslesen
        std::string serverIp = args["--server-ip"];
        int serverPort = std::stoi(args["--server-port"]);
        std::string token = args["--token"];
        std::string name = args["--name"];
        std::string email = args["--email"];
        std::string password = args["--password"];
        std::string language = args.count("--language") ? args["--language"] : "en";
        std::string timezone = args.count("--timezone") ? args["--timezone"] : "Europe/Berlin";
        auto roles = parseListArgument(argc, argv, "--roles");
        auto groups = parseListArgument(argc, argv, "--groups");
        bool usePublic = args.count("--usePublic");
        bool activateFtp = args.count("--activateFtp");
        std::string ftpPassword = args.count("--ftpPassword") ? args["--ftpPassword"] : "";

        // JSON-Payload erstellen
        Json::Value payload;
        payload["command"] = "store_user";
        payload["token"] = token;
        payload["arguments"]["name"] = name;
        payload["arguments"]["email"] = email;
        payload["arguments"]["password"] = password;
        payload["arguments"]["language"] = language;
        payload["arguments"]["timezone"] = timezone;
        for (const auto& role : roles) {
            payload["arguments"]["roles"].append(role);
        }
        for (const auto& group : groups) {
            payload["arguments"]["groups"].append(group);
        }
        payload["arguments"]["usePublic"] = usePublic;
        payload["arguments"]["activateFtp"] = activateFtp;
        payload["arguments"]["ftpPassword"] = ftpPassword;

        std::cout << "📤 Sending store user request...\n";

        // Anfrage senden und Antwort erhalten
        std::string response = sendRequest(serverIp, serverPort, payload);

        std::cout << "✔️ Response from server:\n" << response << "\n";
    } catch (const std::exception& e) {
        std::cerr << "❌ ERROR: " << e.what() << "\n";
        return 1;
    }

    return 0;
}
